tools/epoch-sync/src/cli.rs

```diff
@@ -0,0 +1,293 @@
+                let canonical_epoch_sync_data_hash_result =
+                    epoch_sync_info.get_epoch_sync_data_hash();
+
+                if let Ok(calculated_epoch_sync_data_hash) = calculated_epoch_sync_data_hash_result
+                {
+                    if let Ok(Some(canonical_epoch_sync_data_hash)) =
+                        canonical_epoch_sync_data_hash_result
+                    {
+                        if calculated_epoch_sync_data_hash == canonical_epoch_sync_data_hash {
+                            tracing::info!(
+                                "EpochSyncInfo for height {:?} OK",
+                                epoch_sync_info.epoch_info.epoch_height()
+                            );
+                            continue;
+                        }
+                    }
+                }
+                tracing::error!(
+                    "EpochSyncInfo for height {:?} ERROR {:?} {:?}",
+                    epoch_sync_info.epoch_info.epoch_height(),
+                    calculated_epoch_sync_data_hash_result,
+                    canonical_epoch_sync_data_hash_result
+                );
+                num_errors += 1;
+            } else {
+                cur_hash = hash_to_prev_hash[&cur_hash];
+            }
+        }
+        assert_eq!(num_errors, 0);
+        Ok(())
+    }
+}
+
+fn get_hash_to_prev_hash_from_block_info(
+    storage: &NodeStorage,
+) -> anyhow::Result<HashMap<CryptoHash, CryptoHash>> {
+    let mut hash_to_prev_hash = HashMap::new();
+    let store = storage.get_split_store().unwrap_or(storage.get_hot_store());
+    for result in store.iter(DBCol::BlockInfo) {
+        let (_, value) = result?;
+        let block_info =
+            BlockInfo::try_from_slice(value.as_ref()).expect("Failed to deser BlockInfo");
+        if block_info.hash() != block_info.prev_hash() {
+            hash_to_prev_hash.insert(*block_info.hash(), *block_info.prev_hash());
+        }
+    }
+    Ok(hash_to_prev_hash)
+}
+
+fn get_hash_to_prev_hash_from_block_header(
+    storage: &NodeStorage,
+) -> anyhow::Result<HashMap<CryptoHash, CryptoHash>> {
+    let mut hash_to_prev_hash = HashMap::new();
+    for result in storage.get_hot_store().iter(DBCol::BlockHeader) {
+        let (_, value) = result?;
+        let block_header =
+            BlockHeader::try_from_slice(value.as_ref()).expect("Failed to deser BlockHeader");
+        if block_header.hash() != block_header.prev_hash() {
+            hash_to_prev_hash.insert(*block_header.hash(), *block_header.prev_hash());
+        }
+    }
+    Ok(hash_to_prev_hash)
+}
+
+fn get_hash_to_next_hash(
+    hash_to_prev_hash: &HashMap<CryptoHash, CryptoHash>,
+    chain_head: &CryptoHash,
+) -> anyhow::Result<HashMap<CryptoHash, CryptoHash>> {
+    let mut hash_to_next_hash = HashMap::new();
+    let mut cur_head = *chain_head;
+    while let Some(prev_hash) = hash_to_prev_hash.get(&cur_head) {
+        hash_to_next_hash.insert(*prev_hash, cur_head);
```

Comment:
You can `assert!()` that the insertion happened. This would detect forks in the recent epochs.

Reply:
So, with forks, the mapping from `cur_hash` to `next_hash` is not correct, as we may encounter key collision.
To avoid that I don't build this mapping from the whole `hash_to_prev_hash` mapping, but only from one chain that end with provided `chain_head`.

Reply:
Please add this as a comment to the code.